README.md

# Automation Tycoon
A mod for the game Factorio

## Short description
Brings concepts of a famous Tycoon game into Factorio.


## Long description

You've found out that your belts are not waterproof and cannot be built
outside of buildings.
You have to use your engineering skills to connect your factory buildings
together. Through, technology, you will start transporting your items using
Haulers, trains, robots, cargo ships and cargo planes. Multimodal transport yeah !

[Car transport](doc/base.png)


For researching technologies, you have to deal with an ancient passive
civilization who has established cities on the planet.
They are energy spirits only so they cannot gather resource for themselves.
Provide what they need, and they
will give you some ucoins (the planet currency) in exchange.
Spend ucoins to unlock technologies.

You can't miss the Ancient' cities as they look like big assembling
machines, duh.

[Ancient city](doc/city_outpost.png)

## Get started

Don't try this mod in an save game. You will get frustrated.
Better to start with a new game!
Rethink how to build a factory without belts.

IMPORTANT :
In Factorissimo2 settings, disable Factorissimo2 recursion tech and
set the research to easy.

## Settings

If you feel you miss your belts, you can re-enable them in the settings.


## Know bugs

- Map generation settings does not work for cities placement yet.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.9.0
  Features:
    - You cannot build neither belts outside as they are not waterproof (duh).
    You can only build them in Factorissimo2 buildings.
    - A Factorissimo2 level 0 is now available when starting a game.
    - Cities spawn in the map randomly. Providing what they need gives you cash in return.
    - [TechnologyMarket] Spend your cash to research new technologies. Labs now accepts ucoin.
    - As there is no more belts outside, mining-drills are not easy to automate at low-tech. You might have to
     use AAI miners instead.
    - Get better at Multimodal transport or combined transport:
      - Use AAI Hauler to transport items between buildings.
      - Use Train to transport items between buildings.
      - Use Air to transport items between buildings. (v1.0.0)
    - Of course, native are not going to help you in the task.


  Changes:
    None


city.lua
-- Big Assembly prototype and item definition
local table = require('__stdlib__/stdlib/utils/table')
require("adjustVisuals")
require("util")

data:extend({
  {
    type = "autoplace-control",
    category = "resource",
    name = "at-city"
  },
  {
  name = "cities",
  type = "noise-layer"
  }
})

local function city_autoplace_settings(multiplier, rectangle)
  local peak =
  {
    noise_layer = "cities",
    noise_octaves_difference = -2,
    noise_persistence = 0.9
  }

  if rectangle ~= nil then
    local aux_center = (rectangle[2][1] + rectangle[1][1]) / 2
    local aux_range = math.abs(rectangle[2][1] - rectangle[1][1]) / 2
    local water_center = (rectangle[2][2] + rectangle[1][2]) / 2
    local water_range = math.abs(rectangle[2][2] - rectangle[1][2]) / 2

    peak["aux_optimal"] = aux_center
    peak["aux_range"] = aux_range
    peak["aux_max_range"] = water_range + 0.05

    peak["water_optimal"] = water_center
    peak["water_range"] = water_range
    peak["water_max_range"] = water_range + 0.05
  end

  return
  {
    order = "city",
    coverage = multiplier * 0.01,
    sharpness = 0.7,
    max_probability = multiplier * 0.7,
    peaks = { peak }
  }
end

local function commonAdjustments(factory)
    factory.minable = nil
    factory.next_upgrade = nil
    factory.fast_replaceable_group = nil
    factory.dying_explosion = "big-explosion"
    factory.max_health = 1600

    factory.scale_entity_info_icon = true
    factory.autoplace = city_autoplace_settings(0.085, {{1, 1}, {1, 1}})
    factory.create_ghost_on_death = false
    factory.flags = {
      -- "indestructible",
      -- "not-minable",
      "placeable-neutral", "not-deconstructable", "not-blueprintable",
      "not-rotatable", "not-repairable", "not-flammable", "hide-alt-info"
    }
    factory.collision_mask = factory.collision_mask or {"item-layer", "object-layer", "player-layer", "water-tile"}
    table.insert(factory.collision_mask, "resource-layer")
    factory.resistances = {{type="poison", percent=90}, {type="acid", percent=80}, {type="physical", percent=70},
        {type="fire", percent=70}, {type="explosion", percent=-100}}

    factory.has_backer_name = nil
end

local function create_city()
    local name = "at-city"
    local city = table.deep_copy(data.raw["assembling-machine"]["assembling-machine-3"])
    local icon = "__AutomationTycoon__/graphics/icons/big-assembly.png"

    city.name = name
    city.icon = icon
    city.localised_name = "City"

    city.collision_box = {{-8.1, -8.1}, {8.1, 8.1}}
    city.selection_box = {{-8.8, -9}, {8.8, 9}}
    city.drawing_box = {{-8.8, -8.8}, {8.8, 8.8}}

    city.crafting_categories = {"at-ucoin"}
    city.crafting_speed = 1
    city.order = "a"

    -- city.energy_usage =
    city.ingredient_count = 10
    city.module_specification.module_slots = 5
    city.map_color = {r=103, g=247, b=247}

    city.energy_source = {
      type = "void",
      render_no_power_icon = false,
      render_no_network_icon = false
    }

    city.fixed_recipe = "at-".. "ucoin" .."-to-ucoin"

    commonAdjustments(city)

    local function fluidBox(type, position)
        retvalue = {
                production_type = type,
                pipe_picture = assembler3pipepictures(),
                pipe_covers = pipecoverspictures(),
                base_area = 10,
                pipe_connections = {{ type=type, position=position }},
                secondary_draw_orders = { north = -1 }
            }
        if type == "input" then
            retvalue.base_level = -1
        else
            retvalue.base_level = 1
        end
        return retvalue
    end

    city.fluid_boxes = {}
        -- city.fluid_boxes = {
        --     fluidBox("input", {0, -9}),
        --     fluidBox("input", {-9, 0}),
        --     fluidBox("output", {9, 0}),
        --     fluidBox("output", {0, 9}),
        -- }

    adjustVisuals(city, 6, 1/32)

    data.raw["assembling-machine"][name] = city
end

create_city()


control.lua
require("utils")
local table = require('__stdlib__/stdlib/utils/table')
local Event = require('__stdlib__/stdlib/event/event')
local Force = require('__stdlib__/stdlib/event/force').register_events()

Event.protected_mode = true

if settings.global.forbid_building_belts_on_nauvis then
	local forbidden_entities_on_nauvis = table.array_to_dictionary({
		"transport-belt",
		"underground-belt",
		"splitter",
		--	"inserter"
	}, true)

	Event.register({defines.events.on_built_entity, defines.events.on_robot_built_entity}, function (event)
		-- called at player creation
		local created_entity = event.created_entity
		-- LOG("created_entity.type=" .. created_entity.type)
		if forbidden_entities_on_nauvis[created_entity.type] and not string.find(created_entity.surface.name, "^Factory f") then
			created_entity.surface.create_entity{name="flying-text", position=created_entity.position, text="Cannot build belts on Nauvis"}
			created_entity.destroy()
		end
	end)

end

Event.register({"on_init", "on_configuration_changed"}, function()
	LOG("on_init/on_configuration_changed")
	local data = global
	local ucoin_recipes = {}

	table.each(game.recipe_prototypes, function(recipe)
		if not recipe.enabled then
			return
		end
		local f = recipe.name ~= nil and #recipe.name >= 3 and
			string.sub(recipe.name, 1, 3) == "at-"
		if f then
			-- activate default recipe (resource for instance)
			local ucoin_recipe_name = recipe.name
			LOG("unlocked ucoin recipe is now available ucoin_recipe_name=" .. ucoin_recipe_name)
			table.insert(ucoin_recipes, ucoin_recipe_name)
		else
			table.each(recipe.products, function(product)
				if product.type ~= "item" then
					return
				end
				LOG("product.name=" .. product.name)
				local ucoin_recipe_name = "at-".. product.name .."-to-ucoin"
				local ucoin_recipe = game.recipe_prototypes[ucoin_recipe_name]
				if ucoin_recipe ~= nil then
					LOG("ucoin recipe is now available ucoin_recipe_name=" .. ucoin_recipe_name)
					table.insert(ucoin_recipes, ucoin_recipe_name)
				end
			end) -- end table.each
		end
	end)

	-- local _, force_data = Force.get(technology.force)
	local data = global
	local g_ucoin_recipes_index = 1
	local g_ucoin_recipes = {}
	if #ucoin_recipes > 0 then
		g_ucoin_recipes[g_ucoin_recipes_index] = ucoin_recipes
		g_ucoin_recipes_index = g_ucoin_recipes_index + 1
	end
	data.ucoin_recipes = g_ucoin_recipes
	data.ucoin_recipes_index = g_ucoin_recipes_index
	LOG("g_ucoin_recipes_index=" .. g_ucoin_recipes_index)
	LOG("g_ucoin_recipes=" .. serpent.block(g_ucoin_recipes))

end)

Event.register(defines.events.on_research_finished, function (event)
	local technology = event.research
	LOG("on_research_finished technology.name=" .. technology.name)
	local ucoin_recipes = {}

	if technology.effects ~= nil then
		table.each(technology.effects, function(modifier)
			if modifier.type ~= "unlock-recipe" then
				return
			end
			LOG("modifier.recipe=" .. modifier.recipe)
			local recipe = game.recipe_prototypes[modifier.recipe]
			if recipe ~= nil then
				table.each(recipe.products, function(product)
					if product.type ~= "item" then
						return
					end
					LOG("product.name=" .. product.name)
					local ucoin_recipe_name = "at-".. product.name .."-to-ucoin"
					local ucoin_recipe = game.recipe_prototypes[ucoin_recipe_name]
					if ucoin_recipe ~= nil then
						LOG("ucoin recipe is now available ucoin_recipe_name=" .. ucoin_recipe_name)
						table.insert(ucoin_recipes, ucoin_recipe_name)
					end
				end) -- end table.each
			end -- if recipe ~= nil
		end)
	end

	if #ucoin_recipes > 0 then
		-- local _, force_data = Force.get(technology.force)
		local data = global
		local g_ucoin_recipes_index = data.ucoin_recipes_index or 1
		local g_ucoin_recipes = data.ucoin_recipes or {}
		g_ucoin_recipes[g_ucoin_recipes_index] = ucoin_recipes
		data.ucoin_recipes = g_ucoin_recipes
		data.ucoin_recipes_index = g_ucoin_recipes_index + 1
		LOG("g_ucoin_recipes_index=" .. g_ucoin_recipes_index)
	end
end)


Event.register(defines.events.on_chunk_generated, function (event)
	local surface = event.surface
	LOG("on_chunk_generated")
	table.each(surface.find_entities_filtered {
		area = event.area,
		name = "at-city"
	}, function(city)
		assert(city ~= nil)
		LOG("position" .. serpent.block(city.position))
		local former_recipe = city.get_recipe()
		if former_recipe ~= nil and former_recipe.name ~= "at-ucoin-to-ucoin" then
			LOG("recipe already set, skipping. recipe=" .. former_recipe.name)
			return
		end
		local data = global
		local g_ucoin_recipes_index = data.ucoin_recipes_index or 1
		local g_ucoin_recipes = data.ucoin_recipes or {}
		local window_size = 4

		local min_i = g_ucoin_recipes_index - window_size
		min_i = min_i > 1 and min_i or 1
		local max_i = g_ucoin_recipes_index + window_size
		max_i = max_i <= #g_ucoin_recipes and max_i or #g_ucoin_recipes

		LOG("#g_ucoin_recipes=" .. #g_ucoin_recipes .. " min_i=" .. min_i .. " max_i=" .. max_i)
		local s = table.flatten(table.slice(g_ucoin_recipes, min_i, max_i))
		LOG("#s=" .. #s)
		if #s > 0 then
			local i = math.floor(math.random() * #s) + 1
			LOG("i=" .. i)
			LOG("s[i]=" .. s[i])
			city.set_recipe(s[i])
		end


	end)

end)


data-final-fixes.lua
require("utils")
local table = require('__stdlib__/stdlib/utils/table')

-- -- hide mining-drills
-- local entities_to_hide = table.array_to_dictionary(table.map(data.raw["mining-drill"], function(entity, name)
-- 	return entity.name
-- end), true)
--
-- table.each(data.raw.recipe, function(recipe, name)
-- 	if entities_to_hide[recipe.result] or recipe.results and table.any(recipe.results, function (result)
-- 			return result.type == "item" and entities_to_hide[result.name]
-- 	end) then
-- 		recipe.hidden = true
-- 	end
-- end)

-- require("utils")
-- local table = require('__stdlib__/stdlib/utils/table')
require("bigassembly")

data:extend({
	{
    type = "recipe-category",
    name = "at-ucoin"
	}
})

local function create_ucoin_recipe(item_name, price, enabled)
	LOG("create_ucoin_recipe item_name=" .. item_name .. " price=" .. price)
	data:extend({
		{
			type = "recipe",
			name = "at-".. item_name .."-to-ucoin",
			category = "at-ucoin",
			icon = "__LibMoney__/graphics/ucoin.png",
			icon_size = 32,
			enabled = enabled or false,
			hidden = true,
			energy_required = 1,
			ingredients = {
				{item_name, 1}
			},
			results = {
				{
					name = "ucoin",
					amount = price
				},
			}
		}
	})
end

vanilla_resources_prices = {
--	["water"] = 0,
	["coal"] = 16,
	["stone"] = 27,
	["iron-ore"] = 19,
	["copper-ore"] = 21,
--	["crude-oil"] = 100,
}

local cached_technology_cost = {}
local item_cost = {}

local technology_time_unit_cost = 5
local technology_complexity_cost = 2
local technology_count_cost = 1

local recipe_energy_required_cost = 7
local recipe_complexity_cost = 2

local function get_techonology_cost(technology)
	assert(technology ~= nil)
	LOG("get_techonology_cost technology.name=" .. technology.name)

	local cached_cost = cached_technology_cost[technology.name]
	if cached_cost ~= nil then
		return cached_cost
	end

	local local_cost =
		(technology.unit.count ~= nil and technology.unit.count or 0) * technology_count_cost +
		#technology.unit.ingredients * technology_complexity_cost +
		technology.unit.time * technology_time_unit_cost

	local deps_cost = 0
	if technology.prerequisites ~= nil then
		local deps_cost = table.sum(table.map(technology.prerequisites, function(prereq_technology_name)
			assert (prereq_technology_name ~= nil)
			local prereq_technology = data.raw.technology[prereq_technology_name]
			if prereq_technology == nil then
				LOG("Prerequisited technology \"" .. prereq_technology_name .. " does not exist.")
				return 0
			else
				return get_techonology_cost(prereq_technology)
			end
		end))
	end
	local c = deps_cost + local_cost
	cached_technology_cost[technology.name] = c
	return c
end

local recipe_to_technology = {}
local technology_index = 1
local all_technologies = table.keys(data.raw.technology, true)

-- LOG(serpent.block(all_technologies))

local function get_unlocking_techonology_from_recipe(recipe_name)
	LOG("get_unlocking_techonology_from_recipe recipe_name=" .. recipe_name .. " technology_index=" .. technology_index)

	local saved = recipe_to_technology[recipe_name]
	if saved ~= nil then
		return saved
	end

	-- LOG("#data.raw.technology=" .. #all_technologies)
	while technology_index <= #all_technologies do
		local technology = data.raw.technology[all_technologies[technology_index]]
		LOG("technology.name=" .. technology.name)
		local found = nil
		if technology.effects ~= nil then
			table.each(technology.effects, function(effect)
				if effect.type ~= "unlock-recipe" then
					return
				end
				if not found and effect.recipe == recipe_name then
					found = true
				end
				recipe_to_technology[effect.recipe] = technology
			end)
		end
		technology_index = technology_index + 1
		if found then
			return technology
		end
	end -- while

	return nil
end
-- in
-- recipes -> item
 --> price is recipe + techno

-- out
-- item list with price
--   \-> recipe to ucoin

local function get_recipe_cost_for_result(recipe, result_name)
	-- LOG("get_recipe_cost_for_result recipe.name=" .. recipe.name .. " result_name=" .. result_name)
	local technology = get_unlocking_techonology_from_recipe(recipe.name)
	local technology_cost = 0
	if technology ~= nil then
		technology_cost =  get_techonology_cost(technology)
	end

	local recipe_cost = (function(recipe_sub)
		if #recipe_sub.ingredients == 0 then
			return 0
		end

		local result_count
		if recipe_sub.results ~= nil then
			result_count = table.find(recipe_sub.results, function(v)
				return v.name == result_name
			end).amount
		else
			assert(recipe_sub.result ~= nil)
			assert(result_name == recipe_sub.result)
			result_count = recipe_sub.result_count or 1
		end
		return math.floor(
			((recipe.energy_required or 0.5) * recipe_energy_required_cost +
			#recipe_sub.ingredients * recipe_complexity_cost) / result_count)
	end)(recipe.normal or recipe.expensive or recipe)

	LOG("get_recipe_cost_for_result recipe.name=" .. recipe.name .. " result_name=" .. result_name .." technology_cost=" .. technology_cost .. " recipe_cost=" .. recipe_cost)
	return technology_cost + recipe_cost
end

local item_to_recipes = {}

local create_ucoin_recipes_from_item
do
	local ucoin_recipe_created = {}
	create_ucoin_recipes_from_item = function (item)
		assert(item ~= nil)
		LOG("create_ucoin_recipes_from_item item.name=" .. item.name)

		if ucoin_recipe_created[item.name] ~= nil then
			return
		end

		local recipes = item_to_recipes[item.name]
		assert(recipes ~= nil)
		local recipe_cost = table.min(table.map(recipes, function(recipe)
			return get_recipe_cost_for_result(recipe, item.name)
		end))
		if recipe_cost and recipe_cost > 0 then
			create_ucoin_recipe(item.name, recipe_cost)
			ucoin_recipe_created[item.name] = true
		else
			LOG("No price for item " .. item.name)
		end
	end
end

local function link_recipe_to_item(recipe, item_name)
	assert(item_name ~= nil)

	local i = item_to_recipes[item_name]
	if i ~= nil then
		table.insert(i, recipe)
	else
		item_to_recipes[item_name] = {recipe}
	end
end

local function find_all_results_sub(recipe_sub)
	assert(recipe_sub ~= nil)
	if recipe_sub.result ~= nil then
		return {recipe_sub.result}
	else
		assert(recipe_sub.results ~= nil, "result and results both set for recipe " .. recipe_sub.name)
		return table.map(table.filter(recipe_sub.results, function(i)
			return i.type == "item"
		end), function(i)
			return i.name
		end)
	end
end

local function find_all_results(recipe)
	assert(recipe ~= nil)
	-- TODO: Maybe go through all recipes including difficulties
	-- as output might be differents
	return find_all_results_sub(recipe.normal or recipe.expensive or recipe)
end
-- start from recipe so we know player can craft/assemble the item
table.each(data.raw.recipe, function(recipe)
		table.each(find_all_results(recipe), function(result_name)
			link_recipe_to_item(recipe, result_name)
		end)
end)

LOG(serpent.block(table.map(item_to_recipes, function(value, key)
 	return #value
end)))

table.each(table.keys(item_to_recipes, true), function(item_name)
	local i =
		data.raw.item[item_name] or
		data.raw.tool[item_name] or
		data.raw.armor[item_name] or
		data.raw.gun[item_name] or
		data.raw.module[item_name] or
		data.raw.capsule[item_name] or
		data.raw["item-with-entity-data"][item_name] or
		data.raw["repair-tool"][item_name] or
		data.raw["rail-planner"][item_name] or
		data.raw.ammo[item_name]
	if i == nil then
		LOG("ignoring item " .. item_name)
	else
		create_ucoin_recipes_from_item(i)
	end
end)

table.each(data.raw.resource, function(resource)
	local price = vanilla_resources_prices[resource.name]
	if price == nil then
		LOG("unknown price for resource " .. resource.name)
	else
		create_ucoin_recipe(resource.name, price, true)
	end
end)

create_ucoin_recipe("ucoin", 0)


data-updates.lua
require("utils")
local table = require('__stdlib__/stdlib/utils/table')

-- -- hide mining-drills
-- local entities_to_hide = table.array_to_dictionary(table.map(data.raw["mining-drill"], function(entity, name)
-- 	return entity.name
-- end), true)
--
-- table.each(data.raw.recipe, function(recipe, name)
-- 	if entities_to_hide[recipe.result] or recipe.results and table.any(recipe.results, function (result)
-- 			return result.type == "item" and entities_to_hide[result.name]
-- 	end) then
-- 		recipe.hidden = true
-- 	end
-- end)

-- require("utils")
-- local table = require('__stdlib__/stdlib/utils/table')
require("city")

data:extend({
	{
    type = "recipe-category",
    name = "at-ucoin"
	}
})

local UINT16_MAX = 65535
local function create_ucoin_recipe(item_name, price, enabled)
	LOG("create_ucoin_recipe item_name=" .. item_name .. " price=" .. price)
	local results = {}
	while price > UINT16_MAX do
		table.insert(results, {
			type = "item",
			name = "ucoin",
			amount = UINT16_MAX
		})
		price = price - UINT16_MAX
	end
	table.insert(results, {
		type = "item",
		name = "ucoin",
		amount = price % UINT16_MAX
	})


	data:extend({
		{
			type = "recipe",
			name = "at-".. item_name .."-to-ucoin",
			category = "at-ucoin",
			icon = "__LibMoney__/graphics/ucoin.png",
			icon_size = 32,
			enabled = enabled or false,
			hidden = true,
			energy_required = 1,
			ingredients = {
				{item_name, 1}
			},
			results = results,
			main_product = "ucoin"
		}
	})
end

vanilla_resources_prices = {
--	["water"] = 0,
	["coal"] = 8,
	["coal"] = 16,
	["stone"] = 27,
	["iron-ore"] = 19,
	["copper-ore"] = 21,
	["uranium-ore"] = 71,
--	["crude-oil"] = 100,
}

local cached_technology_cost = {}
local item_cost = {}

local technology_time_unit_cost = 5
local technology_complexity_cost = 2
local technology_count_cost = 1

local recipe_energy_required_cost = 7
local recipe_complexity_cost = 2

local function get_techonology_cost(technology)
	assert(technology ~= nil)
	LOG("get_techonology_cost technology.name=" .. technology.name)

	local cached_cost = cached_technology_cost[technology.name]
	if cached_cost ~= nil then
		return cached_cost
	end

	local local_cost =
		(technology.unit.count ~= nil and technology.unit.count or 0) * technology_count_cost +
		#technology.unit.ingredients * technology_complexity_cost +
		technology.unit.time * technology_time_unit_cost

	local deps_cost = 0
	if technology.prerequisites ~= nil then
		local deps_cost = table.sum(table.map(technology.prerequisites, function(prereq_technology_name)
			assert (prereq_technology_name ~= nil)
			local prereq_technology = data.raw.technology[prereq_technology_name]
			if prereq_technology == nil then
				LOG("Prerequisited technology \"" .. prereq_technology_name .. " does not exist.")
				return 0
			else
				return get_techonology_cost(prereq_technology)
			end
		end))
	end
	local c = deps_cost + local_cost
	cached_technology_cost[technology.name] = c
	return c
end

local recipe_to_technology = {}
local technology_index = 1
local all_technologies = table.keys(data.raw.technology, true)

-- LOG(serpent.block(all_technologies))

local function get_unlocking_techonology_from_recipe(recipe_name)
	-- LOG("get_unlocking_techonology_from_recipe recipe_name=" .. recipe_name .. " technology_index=" .. technology_index)

	local saved = recipe_to_technology[recipe_name]
	if saved ~= nil then
		return saved
	end

	-- LOG("#data.raw.technology=" .. #all_technologies)
	while technology_index <= #all_technologies do
		local technology = data.raw.technology[all_technologies[technology_index]]
		-- LOG("technology.name=" .. technology.name)
		local found = nil
		if technology.effects ~= nil then
			table.each(technology.effects, function(effect)
				if effect.type ~= "unlock-recipe" then
					return
				end
				if not found and effect.recipe == recipe_name then
					found = true
				end
				recipe_to_technology[effect.recipe] = technology
			end)
		end
		technology_index = technology_index + 1
		if found then
			return technology
		end
	end -- while

	return nil
end
-- in
-- recipes -> item
 --> price is recipe + techno

-- out
-- item list with price
--   \-> recipe to ucoin

local function get_recipe_cost_for_result(recipe, result_name)
	-- LOG("get_recipe_cost_for_result recipe.name=" .. recipe.name .. " result_name=" .. result_name)
	local technology = get_unlocking_techonology_from_recipe(recipe.name)
	local technology_cost = 0
	if technology ~= nil then
		technology_cost =  get_techonology_cost(technology)
	end

	local recipe_cost = (function(recipe_sub)
		if #recipe_sub.ingredients == 0 then
			return 0
		end

		local result_count
		if recipe_sub.results ~= nil then
			result_count = table.find(recipe_sub.results, function(v)
				return v.name == result_name
			end).amount
		else
			assert(recipe_sub.result ~= nil)
			assert(result_name == recipe_sub.result)
			result_count = recipe_sub.result_count or 1
		end
		return math.floor(
			((recipe.energy_required or 0.5) * recipe_energy_required_cost +
			#recipe_sub.ingredients * recipe_complexity_cost) / result_count)
	end)(recipe.normal or recipe.expensive or recipe)

	-- LOG("get_recipe_cost_for_result recipe.name=" .. recipe.name .. " result_name=" .. result_name .." technology_cost=" .. technology_cost .. " recipe_cost=" .. recipe_cost)
	return technology_cost + recipe_cost
end

local item_to_recipes = {}

local create_ucoin_recipes_from_item
do
	local ucoin_recipe_created = {}
	create_ucoin_recipes_from_item = function (item)
		assert(item ~= nil)
		-- LOG("create_ucoin_recipes_from_item item.name=" .. item.name)

		if ucoin_recipe_created[item.name] ~= nil then
			return
		end

		local recipes = item_to_recipes[item.name]
		assert(recipes ~= nil)
		local recipe_cost = table.min(table.map(recipes, function(recipe)
			return get_recipe_cost_for_result(recipe, item.name)
		end))
		if recipe_cost and recipe_cost > 0 then
			create_ucoin_recipe(item.name, recipe_cost)
			ucoin_recipe_created[item.name] = true
		else
			LOG("No price for item " .. item.name)
		end
	end
end

local function link_recipe_to_item(recipe, item_name)
	assert(item_name ~= nil)

	local i = item_to_recipes[item_name]
	if i ~= nil then
		table.insert(i, recipe)
	else
		item_to_recipes[item_name] = {recipe}
	end
end

local function find_all_results_sub(recipe_sub)
	assert(recipe_sub ~= nil)
	if recipe_sub.result ~= nil then
		return {recipe_sub.result}
	elseif recipe_sub.results ~= nil then
		return table.map(table.filter(recipe_sub.results, function(i)
			return i.type == "item"
		end), function(i)
			return i.name
		end)
	end
end

local function find_all_results(recipe)
	assert(recipe ~= nil)
	-- TODO: Maybe go through all recipes including difficulties
	-- as output might be differents
	local ret = find_all_results_sub(recipe.normal or recipe.expensive or recipe)
	if #ret == 0 then
		LOG("not results found for recipe " .. recipe.name)
	end
	return ret
end
-- start from recipe so we know player can craft/assemble the item
table.each(data.raw.recipe, function(recipe)
		table.each(find_all_results(recipe), function(result_name)
			link_recipe_to_item(recipe, result_name)
		end)
end)

LOG(serpent.block(table.map(item_to_recipes, function(value, key)
 	return #value
end)))

table.each(table.keys(item_to_recipes, true), function(item_name)
	local i =
		data.raw.item[item_name] or
		data.raw.tool[item_name] or
		data.raw.armor[item_name] or
		data.raw.gun[item_name] or
		data.raw.module[item_name] or
		data.raw.capsule[item_name] or
		data.raw["item-with-entity-data"][item_name] or
		data.raw["repair-tool"][item_name] or
		data.raw["rail-planner"][item_name] or
		data.raw.ammo[item_name]
	if i == nil then
		LOG("ignoring item " .. item_name)
	else
		create_ucoin_recipes_from_item(i)
	end
end)

table.each(data.raw.resource, function(resource)
	local price = vanilla_resources_prices[resource.name]
	if price == nil then
		LOG("unknown price for resource " .. resource.name)
	else
		create_ucoin_recipe(resource.name, price, true)
	end
end)

-- dummy recipe
create_ucoin_recipe("ucoin", 0)
